apps/web/lib/hooks/useSuperfetch.ts

```
@@ -0,0 +1,38 @@
+import { useSession } from "next-auth/react";
+
+interface SuperOptions extends Omit<RequestInit, "headers"|"body"> {
+  body: object;
+}
+
+export default function useSuperfetch(
+  url: string,
+  //disable ts suggestion for headers since we are setting them up manually and allow body to be object
+  options?: SuperOptions
+) {
+  const { data: session } = useSession();
+
+  const superfetch = () => {
+    const headers: HeadersInit = {
+      "Content-Type": "application/json",
+      accept: "application/json",
+      Authorization: "Bearer " + session?.user.accessToken,
+    };
+
+    return fetch(url, {
+      ...options,
+      body: JSON.stringify(options?.body),
+      headers,
+    })
+      .then(async (res) => {
+        if (res.ok) {
+          return { data: await res.json(), status: res.status };
+        }
+        return { data: null, status: res.status };
```

@bdeluga wouldn't it be better to throw an error here as well, so `superfetch` consumers could handle all problems in `onError` instead of slitting it between `onSuccess` and `onError`? @wlechowicz what do you think?

---

That would be better indeed, but JavaScript expects an error as a string. Unless you want me to pass the status as a string, I would have to create a CustomError to pass an object to it. Then, you would destructure the status anyway (due to the context of status throughout the whole app), which would be the same as casting the status string to a number.
### So...
It's up to the Team if we want to check for empty data to show some error. @intive/patronage-2023-js

---

That's a bit awkward, since react-query already provides `{ data, error, isLoading, isError }` this should leverage that.
You shouldn't expect people to look at if !isError and then look again what status does `data` contain to make sure if it really is not an error.
If `!res.ok` then query should return `error`.
> The promise that is returned should either resolve the data or throw an error. For React Query to determine a query has errored, the query function **must** throw. Any error that is thrown in the query function will be persisted on the error state of the query.

https://tanstack.com/query/v3/docs/react/guides/query-functions#usage-with-fetch-and-other-clients-that-do-not-throw-by-default

I supposed for out use case, it would suffice to throw the status as a string, or indeed extend the Error class to have both the message, as well as the numerical error.

---

I was thinking about throwing a new error object with status and message (since these two are potentially useful when handling errors), but even throwing status would be good!